providers/implementations/keymgmt/ec_kmgmt.c

```diff
@@ -0,0 +1,726 @@
+                                          pub_key, pub_key_len))
+        goto err;
+
+    if (priv_key != NULL && include_private) {
+        size_t sz;
+        int ecbits;
+
+        /*
+         * Key import/export should never leak the bit length of the secret
+         * scalar in the key.
+         *
+         * For this reason, on export we use padded BIGNUMs with fixed length.
+         *
+         * When importing we also should make sure that, even if short lived,
+         * the newly created BIGNUM is marked with the BN_FLG_CONSTTIME flag as
+         * soon as possible, so that any processing of this BIGNUM might opt for
+         * constant time implementations in the backend.
+         *
+         * Setting the BN_FLG_CONSTTIME flag alone is never enough, we also have
+         * to preallocate the BIGNUM internal buffer to a fixed public size big
+         * enough that operations performed during the processing never trigger
+         * a realloc which would leak the size of the scalar through memory
+         * accesses.
+         *
+         * Fixed Length
+         * ------------
+         *
+         * The order of the large prime subgroup of the curve is our choice for
+         * a fixed public size, as that is generally the upper bound for
+         * generating a private key in EC cryptosystems and should fit all valid
+         * secret scalars.
+         *
+         * For padding on export we just use the bit length of the order
+         * converted to bytes (rounding up).
+         *
+         * For preallocating the BIGNUM storage we look at the number of "words"
+         * required for the internal representation of the order, and we
+         * preallocate 2 extra "words" in case any of the subsequent processing
+         * might temporarily overflow the order length.
+         */
+        ecbits = EC_GROUP_order_bits(ecg);
+        if (ecbits <= 0)
+            goto err;
+        sz = (ecbits + 7 ) / 8;
+        if (!ossl_param_bld_push_BN_pad(tmpl,
+                                        OSSL_PKEY_PARAM_PRIV_KEY,
+                                        priv_key, sz))
+            goto err;
+    }
+
+    ret = 1;
+
+ err:
+    OPENSSL_free(pub_key);
+    return ret;
+}
+
+static ossl_inline
+int ec_set_param_ecdh_cofactor_mode(EC_KEY *ec, const OSSL_PARAM *p)
+{
+    const EC_GROUP *ecg = EC_KEY_get0_group(ec);
+    const BIGNUM *cofactor;
+    int mode;
+
+    if (!OSSL_PARAM_get_int(p, &mode))
+        return 0;
+
+    /*
+     * mode can be only 0 for disable, or 1 for enable here.
+     *
+     * This is in contrast with the same parameter on an ECDH EVP_PKEY_CTX that
+     * also supports mode == -1 with the meaning of "reset to the default for
+     * the associated key".
+     */
+    if (mode < 0 || mode > 1)
+        return 0;
+
+    /*
+     * The ECDH cofactor mode flag can be set only on a key object containing a
+     * private key.
+     */
```

Comment:
I'm wondering if we're not being too strict here, but that may depend on if we view the cofactor mode parameter as a historic hack that we want to limit the use of (and that we really would want gone), or if we think it's a useful hack that can be applied to other operations than key exchange.  If we see it as a historic hack, then this limitation seems fine, but if we view it as a useful hack, then we shouldn't have this kind of limitation here.  KEYMGMT is, after all, a *general* interface to manage provider side key, *not* a specific extension of key derivation.

Reply:
The reason why I am putting this limitation here is exactly to be flexible in what other implementations might want to do.

**Our** implementation, currently, can set that flag only on a private key object. No other version of a key object in this provider should ever be asked for this flag nor used to try and set this flag. While technically we wouldn't fail, doing otherwise (e.g. setting it on a public-only key object or on a domain params-only key object) is prone to subtle risks as it might be not honored by the ecdh implementation or might not be retained or propagated.
These quirks of our current default provider implementation come from history, and because enabling or disabling EC CDH has serious security implications, we want to err on the side of safety, at least at this point, where the testing coverage on EC CDH stuff is quite minimal.

Nothing prevents us or the community to propose a PR where all the undertested cases are properly covered, and some of these limitations are lifted.

On the other hand, having these strict constraints in our provider, does not restrict what other providers will look like w.r.t. how to handle this flag.

Regarding "is this a history hack we want to get rid of" vs "this is a key property that could be of use for multiple operations", I belong to the latter.
It currently applies prominently to ECDH only as that is one algorithm where we are multiplying our secret scalar by a point provided externally, but in theory is a countermeasure that vulnerable keys might want to force for any current or future cryptosystem involving a scalar multiplicatio between the secret key and an untrusted point.

Reply:
Also, I forgot to mention, in this particular instance of this check in `ec_set_param_ecdh_cofactor_mode()` we want to fail if this is not a key object on which it is safe to set the flag: if an application requires us to use this key object always in CDH mode it is very likely that there is a very good security reason for it, so ignoring the request if we can't comply without signaling it should not be allowed.

Reply:
> **Our** implementation, currently, can set that flag only on a private key object. 

Really?  If we look at what we have before this PR, you set the flag on the key object with `EC_KEY_set_flags()`.  I see no such limitations there.  Even looking at the code in `crypto/ec/ec_pmeth.c`, I see no such limitation anywhere.

> While technically we wouldn't fail, doing otherwise (e.g. setting it on a public-only key object or on a domain params-only key object) is prone to subtle risks as it might be not honored by the ecdh implementation or might not be retained or propagated.

Er, what?  The cofactor mode is attached to a key object, not to the private half or the public half, so if the cofactor mode is set on a key object that only contains a public key, then an operation that needs the private half will fail anyway, regardless of the cofactor flag.

What this limitation does is assume that the cofactor mode can never be useful for an operation that uses the public half.  This goes against "this is a key property that could be of use for multiple operations", in my view.

I see you talking a lot about ECDH conditions, but please remember that we're talking about the KEYMGMT implementation, not the ECDH implementation.  This implementation shouldn't be limited to the conditions of ECDH, even though it supports a flag that's currently used only by our ECDH implementation.  Do you understand the difference?

Reply:
Generally speaking, across implementations, running EC CDH affects how the private scalar is used (it's multiplied by the cofactor).

Since well before 1.1.1, only the cofactor_mode flag of the private half is ever checked, so setting it on the public half has no effect.

This is what I mean with it being a parameter of a private key, as it constrains the way an EC private key should be used.

Reply:
> I see you talking a lot about ECDH conditions, but please remember that we're talking about the KEYMGMT implementation, not the ECDH implementation. This implementation shouldn't be limited to the conditions of ECDH, even though it supports a flag that's currently used only by our ECDH implementation. Do you understand the difference?

I do understand it.
If we abstract away from the specific EC CDH flag, we have key objects that are handled by our keymgm API.

We determined that a generic key object can include combinations of "domain parameters", "private key", "public key", "other parameters" (where each of the 4 things can be either a single item or a collection of items).

We classify as "other parameters" anything that does not define (by standards or consensus) a "domain" to which the key instance belongs.

Parameters defining RSA_PSS or RSA_OAEP keys match this definition, so does the flag that says "this key must be used in cofactor mode when executing ECDH".

I believe this kind of key parameter, that restricts how this key object can be used in operations that support multiple modes of operation, have a right to be full fledged key parameters.
They are parameters being attached to a key object, constraining its usage, for its whole lifetime, even if they have corresponding "operation parameters" to select some mode of operation: the scope of the key parameter is the lifetime of the key object, the scope of the operation parameter is the (normally short-lived) operation execution.

On the limits imposed here on these kind of parameters:
I fail to see why it is a problem if a provider (ours or external) says that some of these properties are meaningful only if the key object meets certain criteria, e.g. a parameter is meaningful only if this key object includes a "private key" part.

Reply:
OK...  I disagree with you, but gotta admit I'm at my wits' end on this one.

I think you need to document the limitation this implementation sets for this flag.  The documentation currently says nothing on that, it simply says that this flag can be set on a key object.

Reply:
The order of operations might make this check a bad idea here..

Reply:
Ok, the pushback is enough for me to decide to take this change out of this PR and open a separate PR to discuss it properly.

Reply:
See 2054895da5 and a3622f3370 please!